### ISSUE TITLE
Embedded Reber Grammar and LSTM

### ISSUE THREAD

**Issue by holderm** [fchollet/keras#750]
I use embedded reber grammar for experimenting around with the LSTM layer.
Here is my model:

```
model = Sequential()
model.add(LSTM(7, 10, activation='sigmoid', inner_activation='hard_sigmoid', return_sequences=True))
model.add(Dropout(0.2))
model.add(TimeDistributedDense(10, 7))
model.add(Activation('sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='rmsprop')

model.fit(Xtrain, ytrain, nb_epoch=250, batch_size=1,show_accuracy=True, verbose=2)
score = model.evaluate(Xtest, ytest, batch_size=1,show_accuracy=True, verbose=1)
```

My data features are one-hot decoded and  of following shape: (samples,22,7)
I have 1000 samples for training and 250 for testing and the sequence length is limited to 22 while having 7 features (1 for each letter).

During training (model.fit( ... ) ) the accuracy does not go higher than ca. 0.4 while the loss goes down as expected

```
Epoch 250/250
2s - loss: 0.0339 - acc: 0.4344
```

Finally I run

```
print('Test score:', score[0])
print('Test accuracy:', score[1])
```

and observe

```
('Test score:', 0.011034337697834417)
('Test accuracy:', 0.41848484848484852)
```

So is the accuracy not going any higher?

**Comment by qmiwang** [comment]
may be you should adjust the biases of gate units

**Comment by holderm** [comment]
Thanks for the hint. This causes the accuracy to drop down to ca 0.15 ?
Without Dropout, it drops even below 0.1 ??

**Comment by qmiwang** [comment]
Have you used the function 'sequence.pad_sequences' to pad the sequence length to 22 ? I am wondering whether an Embedding layer with parameter mask_zero=True ( Whether or not the input value 0 is a special "padding" value that should be masked out. This is useful for recurrent layers which may take variable length input.) should be added before LSTM layer? I am not sure. Best wishes! I am not familiar with Keras and am trying to do video classification by Keras. It is still confused me that what is the correct format of the sequence data (e.g. shape) to feed into LSTM.

**Comment by qmiwang** [comment]
@holderm Have you get good results of Embedded Reber Grammar ?

**Comment by holderm** [comment]
@qmiwang Thanks for your comments. I think I don't need zero padding here since I preallocate a 3d array
`Xtrain = np.zeros([numoftrainexamples, maxlen, 7]) ` where `maxlen` is determined by the longest sequence in the training set. But please correct me if this is wrong. The embedding (i.e. the translation from letters to features) is already done when I generate the sequence, where each letter present a 1 out of k (where k = 7 in this case) coding.

**Comment by qmiwang** [comment]
@holderm Thanks for your reply! As you explained, each example has shape (maxlen,7). So there would be some rows with value 0 when the real length of example short than maxlen. Right ?. I think this would influence the result of lstm because the inner state of lstm will change with zero input.  We need something to identify padding. Perhaps keras.layers.core,Masking is what we need. masking layer is used to mask an input sequence by using a mask value to identify padding. This layer copies the input to the output layer with identified padding replaced with 0s and creates an output mask in the process. At each timestep, if the values all equal `mask_value`, then the corresponding mask value for the timestep is 0 (skipped), otherwise it is 1. I hope this will help you. And I am so sorry I misunderstood Embedding layer. Best wishes!

**Comment by holderm** [comment]
Yes, it feels like `keras.layers.core.Masking` is what's needed here. Do you have an example how to use it? I'm quite new to Keras (and Theano) hence I appreciate any input :+1:

**Comment by sdustwdd** [comment]
Hi,
I have some questions:
(1) masking layer is used to mask an input sequence by using a mask value to identify padding.
Is it means that now I have some sequences,but they  have different lengths,we can make sequences into identify  length, the output of masking layer is a 3-D matrix,but the identify length is ??

**Comment by qmiwang** [comment]
@holderm I've created a gist, see here:
https://gist.github.com/qmiwang/1a4fae904c5c8e291107

I saw the source code of keras this morning. there are some clues in models.py. Class Sequential has method 'complie' which gets the mask recursively.

Through this script in the gist above run correctly, I am not sure about the results. Because sometimes, there are two possible chars can be next char (except the last T or P ), hence the corresponding target has two position with value 1 when local representation is used. To test the correct of prediction, we should compare those two positions. I am not clear about this procedure in keras. If you have any idea, please let you know ! thanks！

**Comment by holderm** [comment]
Update from my site:
```
model = Sequential()
model.add(Masking())
model.add(LSTM(7, 10, activation='sigmoid', inner_activation='hard_sigmoid', return_sequences=True))
model.add(TimeDistributedDense(10, 7))
model.add(Activation('sigmoid')))
opt='rmsprop'
model.compile(loss='mean_squared_error', optimizer=opt)
model.fit(Xtrain, ytrain, nb_epoch=250, batch_size=10,show_accuracy=True, verbose=2,shuffle = True, validation_split=0.2)
score = model.evaluate(Xtest, ytest, batch_size=10,show_accuracy=True, verbose=1)

print('Test score:', score[0])
print('Test accuracy:', score[1])
```

gives

```
('Test score:', 0.0058113144452354706)
('Test accuracy:', 0.83786666666666676)
```

I tried different settings and optimizers but I could not push the accuracy higher than 0.83 something. Any ideas why is that?

**Comment by qmiwang** [comment]
keras support two kinds of classification evaluation (binary or category) and there is only one position with value 1 in each label vector. However, Embedded Reber Grammar is not a critical classification problem. At each time step, there are two (or one sometimes) possible chars can be next char , hence the corresponding target has two position with value 1 when local representation is used. Keras can not evaluate this correctly.  Te check the test accuracy, you can use model.predict to get the output of the model and then write a check function by yourself.  By the way, as your results showed, the real accuracy must lager than 0.8378....
